File: bindings/java/src/main/java/systems/zlink/contracts/Poller.java
/* SPDX-License-Identifier: MPL-2.0 */

package systems.zlink.contracts;

import java.lang.foreign.MemorySegment;
import java.time.Duration;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import systems.zlink.contracts.service.spot.Spot;
import systems.zlink.runtime.nativebridge.Native;
import systems.zlink.runtime.nativebridge.RequestProgressPump;

public final class Poller implements AutoCloseable {
    private static final int MASK_POLLCOMPLETION = 32;
    private final List<PollItem> items = new ArrayList<>();
    private MemorySegment handle;

    public Poller() {
        handle = Native.pollerNew();
        if (handle == null || handle.address() == 0)
            throw ZlinkException.fromLastError("zlink_poller_new");
    }

    public void add(Socket socket, long slot, PollEventFlag... events) {
        addSocket(socket, PollEventFlag.combine(events), slot);
    }

    public void add(Spot spot, long slot, PollEventFlag... events) {
        addSpot(spot, PollEventFlag.combine(events), slot);
    }

    public void addFd(int fd, long slot, PollEventFlag... events) {
        ensureOpen();
        validateSlot(slot);
        PollItem item = PollItem.fd(fd, PollEventFlag.combine(events), slot);
        int rc = Native.pollerAddFd(handle, fd, item.userData(), item.events);
        if (rc != 0)
            throw ZlinkException.fromLastError("zlink_poller_add_fd");
        items.add(item);
    }

    public void add(Timer timer, long slot) {
        ensureOpen();
        Objects.requireNonNull(timer, "timer");
        validateSlot(slot);
        PollItem item = PollItem.timer(timer.handle(), slot);
        int rc = Native.pollerAddTimer(handle, timer.handle(), item.userData());
        if (rc != 0)
            throw ZlinkException.fromLastError("zlink_poller_add_timer");
        items.add(item);
    }

    public void modify(Socket socket, PollEventFlag... events) {
        ensureOpen();
        Objects.requireNonNull(socket, "socket");
        int index = findSocket(socket.handle());
        if (index < 0)
            throw new IllegalArgumentException("socket is not registered");
        int mask = PollEventFlag.combine(events);
        int rc = Native.pollerModify(handle, socket.handle(), mask);
        if (rc != 0)
            throw ZlinkException.fromLastError("zlink_poller_modify");
        items.get(index).events = mask;
    }

    public void modify(Spot spot, PollEventFlag... events) {
        ensureOpen();
        Objects.requireNonNull(spot, "spot");
        int index = findSpot(spot.handleInternal());
        if (index < 0)
            throw new IllegalArgumentException("spot is not registered");
        int mask = PollEventFlag.combine(events);
        int rc = Native.pollerModifySpotSub(handle, spot.handleInternal(),
            mask);
        if (rc != 0)
            throw ZlinkException.fromLastError("zlink_poller_modify");
        PollItem item = items.get(index);
        unregisterExternalProgress(item);
        item.events = mask;
        registerExternalProgress(item);
    }

    public void modifyFd(int fd, PollEventFlag... events) {
        ensureOpen();
        int index = findFd(fd);
        if (index < 0)
            throw new IllegalArgumentException("fd is not registered");
        int mask = PollEventFlag.combine(events);
        int rc = Native.pollerModifyFd(handle, fd, mask);
        if (rc != 0)
            throw ZlinkException.fromLastError("zlink_poller_modify_fd");
        items.get(index).events = mask;
    }

    public boolean remove(Socket socket) {
        ensureOpen();
        Objects.requireNonNull(socket, "socket");
        int index = findSocket(socket.handle());
        if (index < 0)
            return false;
        int rc = Native.pollerRemove(handle, socket.handle());
        if (rc != 0)
            throw ZlinkException.fromLastError("zlink_poller_remove");
        unregisterExternalProgress(items.remove(index));
        return true;
    }

    public boolean remove(Spot spot) {
        ensureOpen();
        Objects.requireNonNull(spot, "spot");
        int index = findSpot(spot.handleInternal());
        if (index < 0)
            return false;
        int rc = Native.pollerRemoveSpotSub(handle, spot.handleInternal());
        if (rc != 0)
            throw ZlinkException.fromLastError("zlink_poller_remove");
        unregisterExternalProgress(items.remove(index));
        return true;
    }

    public boolean removeFd(int fd) {
        ensureOpen();
        int index = findFd(fd);
        if (index < 0)
            return false;
        int rc = Native.pollerRemoveFd(handle, fd);
        if (rc != 0)
            throw ZlinkException.fromLastError("zlink_poller_remove_fd");
        unregisterExternalProgress(items.remove(index));
        return true;
    }

    public boolean remove(Timer timer) {
        ensureOpen();
        Objects.requireNonNull(timer, "timer");
        int index = findTimer(timer.handle());
        if (index < 0)
            return false;
        int rc = Native.pollerRemoveTimer(handle, timer.handle());
        if (rc != 0)
            throw ZlinkException.fromLastError("zlink_poller_remove_timer");
        unregisterExternalProgress(items.remove(index));
        return true;
    }

    public void clear() {
        ensureOpen();
        int rc = Native.pollerDestroy(handle);
        if (rc != 0)
            throw ZlinkException.fromLastError("zlink_poller_destroy");
        handle = Native.pollerNew();
        if (handle == null || handle.address() == 0)
            throw ZlinkException.fromLastError("zlink_poller_new");
        unregisterAllExternalProgress();
        items.clear();
    }

    public int size() {
        ensureOpen();
        int rc = Native.pollerSize(handle);
        if (rc < 0)
            throw ZlinkException.fromLastError("zlink_poller_size");
        return rc;
    }

    public int wait(PollEvents events, Duration timeout) {
        ensureOpen();
        Objects.requireNonNull(events, "events");
        if (items.isEmpty()) {
            events.markReadyCount(0);
            return 0;
        }
        int readyCount = Native.pollerWait(handle, events.segment(),
            events.capacity(), toIntMillis(timeout));
        if (readyCount < 0)
            throw ZlinkException.fromLastError("zlink_poller_wait");
        events.markReadyCount(readyCount);
        return readyCount;
    }

    @Override
    public void close() {
        if (handle == null || handle.address() == 0)
            return;
        Native.pollerDestroy(handle);
        handle = MemorySegment.NULL;
        unregisterAllExternalProgress();
        items.clear();
    }

    private void addSocket(Socket socket, int events, long slot) {
        ensureOpen();
        Objects.requireNonNull(socket, "socket");
        validateSlot(slot);
        PollItem item = PollItem.socket(socket.handle(), events, slot);
        int rc = Native.pollerAdd(handle, socket.handle(), item.userData(),
            events);
        if (rc != 0)
            throw ZlinkException.fromLastError("zlink_poller_add");
        items.add(item);
    }

    private void addSpot(Spot spot, int events, long slot) {
        ensureOpen();
        Objects.requireNonNull(spot, "spot");
        validateSlot(slot);
        PollItem item = PollItem.spot(spot.handleInternal(), events, slot);
        int rc = Native.pollerAddSpotSub(handle, spot.handleInternal(),
            item.userData(), events);
        if (rc != 0)
            throw ZlinkException.fromLastError("zlink_poller_add");
        registerExternalProgress(item);
        items.add(item);
    }

    private static void registerExternalProgress(PollItem item) {
        if (item.isSpot && (item.events & MASK_POLLCOMPLETION) != 0) {
            RequestProgressPump.acquireExternalSpotProgress(item.handle);
        }
    }

    private static void unregisterExternalProgress(PollItem item) {
        if (item.isSpot && (item.events & MASK_POLLCOMPLETION) != 0) {
            RequestProgressPump.releaseExternalSpotProgress(item.handle);
        }
    }

    private void unregisterAllExternalProgress() {
        for (PollItem item : items) {
            unregisterExternalProgress(item);
        }
    }

    private void ensureOpen() {
        if (handle == null || handle.address() == 0)
            throw new IllegalStateException("poller is closed");
    }

    private static void validateSlot(long slot) {
        if (slot < 0)
            throw new IllegalArgumentException("slot must be >= 0");
    }

    private static int toIntMillis(Duration timeout) {
        long millis = Objects.requireNonNull(timeout, "timeout").toMillis();
        if (millis < Integer.MIN_VALUE || millis > Integer.MAX_VALUE) {
            throw new IllegalArgumentException(
                "timeout millis out of int range: " + millis);
        }
        return (int) millis;
    }

    private int findSocket(MemorySegment socketHandle) {
        for (int i = 0; i < items.size(); i++) {
            PollItem item = items.get(i);
            if (item.kind == PollSourceKind.SOCKET
                && item.handle.address() == socketHandle.address()) {
                return i;
            }
        }
        return -1;
    }

    private int findFd(int fd) {
        for (int i = 0; i < items.size(); i++) {
            PollItem item = items.get(i);
            if (item.kind == PollSourceKind.FD && item.fd == fd)
                return i;
        }
        return -1;
    }

    private int findSpot(MemorySegment spotHandle) {
        for (int i = 0; i < items.size(); i++) {
            PollItem item = items.get(i);
            if (item.kind == PollSourceKind.SOCKET
                && item.handle.address() == spotHandle.address()) {
                return i;
            }
        }
        return -1;
    }

    private int findTimer(MemorySegment timerHandle) {
        for (int i = 0; i < items.size(); i++) {
            PollItem item = items.get(i);
            if (item.kind == PollSourceKind.TIMER
                && item.handle.address() == timerHandle.address()) {
                return i;
            }
        }
        return -1;
    }

    private static final class PollItem {
        private final PollSourceKind kind;
        private final MemorySegment handle;
        private final int fd;
        private int events;
        private final long slot;
        private final boolean isSpot;

        private PollItem(PollSourceKind kind, MemorySegment handle, int fd,
                         int events, long slot, boolean isSpot) {
            this.kind = kind;
            this.handle = handle;
            this.fd = fd;
            this.events = events;
            this.slot = slot;
            this.isSpot = isSpot;
        }

        static PollItem socket(MemorySegment handle, int events, long slot) {
            return new PollItem(PollSourceKind.SOCKET, handle, 0, events,
                slot, false);
        }

        static PollItem spot(MemorySegment handle, int events, long slot) {
            return new PollItem(PollSourceKind.SOCKET, handle, 0, events,
                slot, true);
        }

        static PollItem fd(int fd, int events, long slot) {
            return new PollItem(PollSourceKind.FD, MemorySegment.NULL, fd,
                events, slot, false);
        }

        static PollItem timer(MemorySegment handle, long slot) {
            return new PollItem(PollSourceKind.TIMER, handle, 0, 0, slot,
                false);
        }

        MemorySegment userData() {
            return MemorySegment.ofAddress(slot);
        }
    }
}

File: bindings/dotnet/src/Zlink/Runtime/Eventing/Poller.cs
// SPDX-License-Identifier: MPL-2.0

using System;
using System.Collections.Generic;
using System.Threading.Tasks;
using Systems.Zlink.Native;

namespace Systems.Zlink;

internal sealed class Poller : IPoller
{
    private static readonly string[] RequiredExports = new[]
    {
        "zlink_poller_new",
        "zlink_poller_destroy",
        "zlink_poller_size",
        "zlink_poller_add",
        "zlink_poller_add_fd",
        "zlink_poller_add_timer",
        "zlink_poller_modify",
        "zlink_poller_modify_fd",
        "zlink_poller_remove",
        "zlink_poller_remove_fd",
        "zlink_poller_remove_timer",
        "zlink_poller_wait"
    };

    private readonly List<PollItem> _items = new();
    private ZlinkPollerEvent[] _nativeEvents = Array.Empty<ZlinkPollerEvent>();
    private IntPtr _handle;

    public Poller()
    {
        List<string> missing = GetMissingExports();
        if (missing.Count > 0)
        {
            throw new ZlinkConfigException(ConfigResult.NotSupported,
                (int)ErrorCode.ENotSup);
        }

        _handle = NativeMethods.zlink_poller_new();
        if (_handle == IntPtr.Zero)
            throw ZlinkException.CreateConfigException(NativeMethods.zlink_errno());
        _items.Clear();
        _nativeEvents = Array.Empty<ZlinkPollerEvent>();
    }

    internal int Count
    {
        get
        {
            EnsureNotDisposed();
            int rc = NativeMethods.zlink_poller_size(_handle, out _);
            if (rc < 0)
                throw ZlinkException.CreateConfigException(
                    NativeMethods.zlink_errno());
            return rc;
        }
    }

    public int Size => Count;

    public void Add(IZlinkSocket socket, PollEventFlags events, nuint slot)
    {
        EnsureNotDisposed();
        IntPtr socketHandle = SocketInterop.RequirePollableHandle(socket,
            nameof(socket));
        EnumValidation.EnsurePollEvents(events, nameof(events));

        IntPtr userData = SlotToUserData(slot);
        int rc = NativeMethods.zlink_poller_add(_handle, socketHandle,
            userData, (short)events);
        if (rc != 0)
            ZlinkException.ThrowConfigIfError(rc);
        RegisterItem(new PollItem(PollItemKind.Socket, socket, socketHandle, 0,
            null, events, slot));
    }

    public void AddFd(int fd, PollEventFlags events, nuint slot)
    {
        EnsureNotDisposed();
        EnumValidation.EnsurePollEvents(events, nameof(events));

        IntPtr userData = SlotToUserData(slot);
        int rc = NativeMethods.zlink_poller_add_fd(_handle, fd, userData,
            (short)events);
        if (rc != 0)
            ZlinkException.ThrowConfigIfError(rc);
        RegisterItem(new PollItem(PollItemKind.Fd, null, IntPtr.Zero, fd, null,
            events, slot));
    }

    public void Add(IZlinkTimer timer, nuint slot)
    {
        EnsureNotDisposed();
        Timer concreteTimer = SocketInterop.RequireTimer(timer, nameof(timer));

        IntPtr userData = SlotToUserData(slot);
        int rc = NativeMethods.zlink_poller_add_timer(_handle,
            concreteTimer.Handle, userData);
        if (rc != 0)
            ZlinkException.ThrowConfigIfError(rc);
        RegisterItem(new PollItem(PollItemKind.Timer, null, IntPtr.Zero, 0,
            concreteTimer, PollEventFlags.PollIn, slot));
    }

    public void Modify(IZlinkSocket socket, PollEventFlags events)
    {
        EnsureNotDisposed();
        IntPtr socketHandle = SocketInterop.RequirePollableHandle(socket,
            nameof(socket));
        EnumValidation.EnsurePollEvents(events, nameof(events));

        int index = FindSocket(socketHandle);
        if (index < 0)
            throw new ArgumentException("socket is not registered",
                nameof(socket));

        int rc = NativeMethods.zlink_poller_modify(_handle, socketHandle,
            (short)events);
        ZlinkException.ThrowConfigIfError(rc);
        UnregisterExternalProgress(_items[index]);
        _items[index].Events = events;
        RegisterExternalProgress(_items[index]);
    }

    public void ModifyFd(int fd, PollEventFlags events)
    {
        EnsureNotDisposed();
        EnumValidation.EnsurePollEvents(events, nameof(events));

        int index = FindFd(fd);
        if (index < 0)
            throw new ArgumentException("fd is not registered", nameof(fd));

        int rc = NativeMethods.zlink_poller_modify_fd(_handle, fd,
            (short)events);
        ZlinkException.ThrowConfigIfError(rc);
        _items[index].Events = events;
    }

    public bool Remove(IZlinkSocket socket)
    {
        EnsureNotDisposed();
        IntPtr socketHandle = SocketInterop.RequirePollableHandle(socket,
            nameof(socket));

        int index = FindSocket(socketHandle);
        if (index < 0)
            return false;

        int rc = NativeMethods.zlink_poller_remove(_handle, socketHandle);
        ZlinkException.ThrowConfigIfError(rc);
        UnregisterItem(index);
        return true;
    }

    public bool Remove(IZlinkTimer timer)
    {
        EnsureNotDisposed();
        Timer concreteTimer = SocketInterop.RequireTimer(timer, nameof(timer));

        int index = FindTimer(concreteTimer.Handle);
        if (index < 0)
            return false;

        int rc = NativeMethods.zlink_poller_remove_timer(_handle,
            concreteTimer.Handle);
        ZlinkException.ThrowConfigIfError(rc);
        UnregisterItem(index);
        return true;
    }

    public bool Remove(int fd)
    {
        EnsureNotDisposed();

        int index = FindFd(fd);
        if (index < 0)
            return false;

        int rc = NativeMethods.zlink_poller_remove_fd(_handle, fd);
        ZlinkException.ThrowConfigIfError(rc);
        UnregisterItem(index);
        return true;
    }

    public void Clear()
    {
        EnsureNotDisposed();

        IntPtr handle = _handle;
        int rc = NativeMethods.zlink_poller_destroy(ref handle);
        ZlinkException.ThrowConfigIfError(rc);

        _handle = NativeMethods.zlink_poller_new();
        if (_handle == IntPtr.Zero)
            throw ZlinkException.CreateConfigException(NativeMethods.zlink_errno());
        UnregisterAllExternalProgress();
        _items.Clear();
        _nativeEvents = Array.Empty<ZlinkPollerEvent>();
    }

    public void Close()
    {
        Dispose();
    }

    public int Wait(Span<PollEvent> destination, TimeSpan timeout)
    {
        EnsureNotDisposed();
        if (destination.Length == 0)
            throw new ArgumentException("destination must not be empty.",
                nameof(destination));
        if (_items.Count == 0)
            return 0;

        int capacity = Math.Min(destination.Length, _items.Count);
        EnsureEventCapacity(capacity);
        int ready = WaitNative(ToTimeoutMilliseconds(timeout), capacity);
        if (ready < 0)
            throw ZlinkException.CreateRecvException(NativeMethods.zlink_errno());
        if (ready == 0)
            return 0;

        for (int i = 0; i < ready; i++)
            destination[i] = MapEvent(_nativeEvents[i]);
        return ready;
    }

    public void Dispose()
    {
        if (_handle == IntPtr.Zero)
            return;

        IntPtr handle = _handle;
        int rc;
        while (true)
        {
            rc = NativeMethods.zlink_poller_destroy(ref handle);
            if (rc == 0)
                break;

            int errno = NativeMethods.zlink_errno();
            ErrorCode code = ZlinkException.MapErrorCode(errno);
            if (code == ErrorCode.EIntr || errno == 4)
                continue;
            break;
        }
        _handle = IntPtr.Zero;
        UnregisterAllExternalProgress();
        _items.Clear();
        _nativeEvents = Array.Empty<ZlinkPollerEvent>();
        if (rc != 0)
            throw ZlinkException.CreateCloseException(NativeMethods.zlink_errno());
        GC.SuppressFinalize(this);
    }

    public ValueTask DisposeAsync()
    {
        Dispose();
        return ValueTask.CompletedTask;
    }

    ~Poller()
    {
        Dispose();
    }

    private static List<string> GetMissingExports()
    {
        var missing = new List<string>();
        foreach (string symbol in RequiredExports)
        {
            if (!NativeLibraryLoader.HasExport(symbol))
                missing.Add(symbol);
        }
        return missing;
    }

    private static int ToTimeoutMilliseconds(TimeSpan timeout)
    {
        if (timeout < TimeSpan.Zero)
            return -1;
        double millis = timeout.TotalMilliseconds;
        if (millis > int.MaxValue)
            return int.MaxValue;
        return (int)Math.Ceiling(millis);
    }

    private void EnsureEventCapacity(int count)
    {
        if (_nativeEvents.Length < count)
            _nativeEvents = new ZlinkPollerEvent[count];
    }

    private unsafe int WaitNative(int timeoutMs, int capacity)
    {
        fixed (ZlinkPollerEvent* events = _nativeEvents)
        {
            return NativeMethods.zlink_poller_wait_pinned(_handle, events,
                capacity, timeoutMs, out _);
        }
    }

    private int FindSocket(IntPtr handle)
    {
        for (int i = 0; i < _items.Count; i++)
        {
            PollItem item = _items[i];
            if (item.IsSocket && item.SocketHandle == handle)
                return i;
        }
        return -1;
    }

    private int FindFd(int fd)
    {
        for (int i = 0; i < _items.Count; i++)
        {
            PollItem item = _items[i];
            if (item.Kind == PollItemKind.Fd && item.Fd == fd)
                return i;
        }
        return -1;
    }

    private int FindTimer(IntPtr handle)
    {
        for (int i = 0; i < _items.Count; i++)
        {
            PollItem item = _items[i];
            if (item.Kind == PollItemKind.Timer && item.Timer?.Handle == handle)
                return i;
        }
        return -1;
    }

    private PollEvent MapEvent(ZlinkPollerEvent nativeEvent)
    {
        PollSourceKind sourceKind = (PollSourceKind)nativeEvent.MonitorSourceKind;
        int fd = sourceKind == PollSourceKind.Fd ? nativeEvent.Fd : 0;
        return new PollEvent(sourceKind, UserDataToSlot(nativeEvent.UserData),
            (PollEventFlags)nativeEvent.Events, fd);
    }

    private static IntPtr SlotToUserData(nuint slot)
    {
        return unchecked((IntPtr)slot);
    }

    private static nuint UserDataToSlot(IntPtr userData)
    {
        return unchecked((nuint)userData);
    }

    private void RegisterItem(PollItem item)
    {
        RegisterExternalProgress(item);
        _items.Add(item);
    }

    private void UnregisterItem(int index)
    {
        PollItem item = _items[index];
        UnregisterExternalProgress(item);
        _items.RemoveAt(index);
    }

    private static void RegisterExternalProgress(PollItem item)
    {
        if (item.IsSocket
            && (item.Events & PollEventFlags.PollCompletion) != 0)
            RequestProgressPump.AcquireExternalProgress(item.SocketHandle);
    }

    private static void UnregisterExternalProgress(PollItem item)
    {
        if (item.IsSocket
            && (item.Events & PollEventFlags.PollCompletion) != 0)
            RequestProgressPump.ReleaseExternalProgress(item.SocketHandle);
    }

    private void UnregisterAllExternalProgress()
    {
        foreach (PollItem item in _items)
            UnregisterExternalProgress(item);
    }

    private void EnsureNotDisposed()
    {
        if (_handle == IntPtr.Zero)
            throw new ObjectDisposedException(nameof(Poller));
    }

    private enum PollItemKind
    {
        Socket,
        Fd,
        Timer
    }

    private sealed class PollItem
    {
        public PollItem(PollItemKind kind, IZlinkSocket? socket,
            IntPtr socketHandle, int fd, Timer? timer, PollEventFlags events,
            nuint slot)
        {
            Kind = kind;
            Socket = socket;
            SocketHandle = socketHandle;
            Fd = fd;
            Timer = timer;
            Events = events;
            Slot = slot;
        }

        public PollItemKind Kind { get; }
        public IZlinkSocket? Socket { get; }
        public IntPtr SocketHandle { get; }
        public int Fd { get; }
        public Timer? Timer { get; }
        public PollEventFlags Events { get; set; }
        public nuint Slot { get; }
        public bool IsSocket => Kind == PollItemKind.Socket;
    }
}
